File: dotnet/Workspace/Resource.cs
using System.Collections.Generic;

namespace Microsoft.Workspace
{
    public class Resource : IResource
    {
        string IResource.Name { get; set; }

        WorkspaceImpl IResource.Workspace { get; set; }

        string[] IResource.Path { get; set; }

        public ICredentialProvider CredentialProvider { get; set; }

        IEnumerable<ICredentialProvider> GetCredentialProviders()
        {
            if (CredentialProvider != null)
            {
                yield return CredentialProvider;
                yield break;
            }

            yield return EnvironmentCredentialProvider.Instance;
        }
    }
}

File: typescript/src/Resource.ts
import Workspace from "./Workspace";

export default class Resource {
	_name: (string | undefined);
	_workspace: (Workspace | undefined);
	_path: (string[] | undefined);

	init(name: string, workspace: Workspace, path: string[]) {
		this._name = name;
		this._workspace = workspace;
		this._path = path;
	}

	get name() {
		return this._name;
	}

	get workspace() {
		return this._workspace;
	}

	get path() {
		return this._path;
	}
}